pkg/operator/ceph/object/user/controller.go

```diff
@@ -18,202 +18,313 @@ limitations under the License.
-		logger.Errorf("failed to create object store user %q. %v", user.Name, err)
-		updateCephObjectStoreUserStatus(user.GetName(), user.GetNamespace(), k8sutil.FailedStatus, c.context)
+	return reconcileResponse, err
+}
 
+func (r *ReconcileObjectStoreUser) reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// Make sure a CephCluster is present otherwise do nothing
+	_, isReadyToReconcile, reconcileResponse := opcontroller.IsReadyToReconcile(r.client, r.context, request.NamespacedName)
+	if !isReadyToReconcile {
+		logger.Debugf("CephCluster resource not ready in namespace %q, retrying in %q.", request.NamespacedName.Namespace, opcontroller.WaitForRequeueIfCephClusterNotReadyAfter.String())
+		return reconcileResponse, nil
 	}
-	updateCephObjectStoreUserStatus(user.GetName(), user.GetNamespace(), k8sutil.ReadyStatus, c.context)
-}
 
-func (c *ObjectStoreUserController) onUpdate(oldObj, newObj interface{}) {
-	if c.clusterSpec.External.Enable && c.clusterSpec.CephVersion.Image == "" {
-		logger.Warningf("Updating object store user for an external ceph cluster is disabled because no Ceph image is specified")
-		return
+	// Fetch the CephObjectStoreUser instance
+	cephObjectStoreUser := &cephv1.CephObjectStoreUser{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, cephObjectStoreUser)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			logger.Debug("CephObjectStoreUser resource not found. Ignoring since object must be deleted.")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, errors.Wrap(err, "failed to get CephObjectStoreUser")
 	}
-	// TODO: Add update code here after features are added which require updates.
-}
 
-func (c *ObjectStoreUserController) onDelete(obj interface{}) {
-	if c.clusterSpec.External.Enable && c.clusterSpec.CephVersion.Image == "" {
-		logger.Warningf("Deleting object store user for an external ceph cluster is disabled because no Ceph image is specified")
-		return
+	// The CR was just created, initializing status fields
+	if cephObjectStoreUser.Status == nil {
+		cephObjectStoreUser.Status = &cephv1.Status{}
+		cephObjectStoreUser.Status.Phase = k8sutil.Created
+		err := opcontroller.UpdateStatus(r.client, cephObjectStoreUser)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to set status")
+		}
 	}
 
-	user, err := getObjectStoreUserObject(obj)
+	// validate the user settings
+	err = ValidateUser(r.context, cephObjectStoreUser)
 	if err != nil {
-		logger.Errorf("failed to get objectstoreuser object. %v", err)
-		return
+		cephObjectStoreUser.Status.Phase = k8sutil.ReconcileFailedStatus
+		err := opcontroller.UpdateStatus(r.client, cephObjectStoreUser)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to set status")
+		}
+		return reconcile.Result{}, errors.Wrapf(err, "invalid pool CR %q spec", cephObjectStoreUser.Name)
```

Comment:
We need to return the original err, right? `err` will always be nil in this code path because the call to `UpdateStatus()` overwrites the original `err`

Reply:
Yes yes I had a comment for this. Thanks.